File: 排序/quicksort.js
function quickSort(arr,left,right){
    if (left===right) return 
    const key=partition(arr,left,right)
    if(left<key){
        quickSort(arr,left,key-1)
    }
    if(right>key){
        quickSort(arr,key+1,right)
    }
}
function partition (arr,left,right) {
  const key=arr[left]
  while(left<right){
      while(key<=arr[right] && left<right){
        right--
      }
      [arr[left], arr[right]] = [arr[right], arr[left]]; // 交换
      while (key >= arr[left] && left < right) {
        // 如果key大于a[left]，则left递增，继续比较
        left++;
      }
      [arr[left], arr[right]] = [arr[right], arr[left]]; // 交换
  }
  return left
}
var quickSort2 = function(arr) {
    　　if (arr.length <= 1) {//如果数组长度小于等于1无需判断直接返回即可 
            return arr;
        }
    　　var pivotIndex = Math.floor(arr.length / 2);//取基准点 
    　　var pivot = arr.splice(pivotIndex, 1)[0];//取基准点的值,splice(index,1)函数可以返回数组中被删除的那个数
    　　var left = [];//存放比基准点小的数组
    　　var right = [];//存放比基准点大的数组 
    　　for (var i = 0; i < arr.length; i++){ //遍历数组，进行判断分配 
    　　　　if (arr[i] < pivot) {
    　　　　　　left.push(arr[i]);//比基准点小的放在左边数组 
    　　　　} else {
    　　　　　　right.push(arr[i]);//比基准点大的放在右边数组 
    　　　　}
    　　}
             //递归执行以上操作,对左右两个数组进行操作，直到数组长度为<=1； 
    　　return quickSort2(left).concat([pivot], quickSort2(right));
    }
const arr = [2, 3, 1, 4];
quickSort(arr, 0, 3);
console.log(arr);
console.log(quickSort2(arr))

File: 排序/quickSort.go

func quickSort(list []int, start, end int) {
    if end-start > 1 {
        // get the pivot
        mid := partition(list, start, end)
        // left sort
        quickSort(list, start, mid)
        // right sort
        quickSort(list, mid+1, end)
    }
}

func partition(list []int, begin, end int) (i int) {
    cValue := list[begin]
    i = begin
    for j := i + 1; j < end; j++ {
        if list[j] < cValue {
            i++ // 这里一定要先加1后在交换值，因为支点现在不必交换，现在i 和 j（小于支点和大于支点）正在划分边界
            list[j], list[i] = list[i], list[j]
            fmt.Println("list:", list)
        }
    }
/* 到此，i和j的边界已经划分完成，把i对应的值和支点对应的值交换后，就符合了快分的要求：i左边对应的值都小于等于且右边的都大于支点对应的值
此时i的值就是新的支点, 对应的值就是新的主元
*/
    list[i], list[begin] = list[begin], list[i]
    return i
}